### PR TITLE
refactor(TimezoneSelector): simplify override logics and tests

### PR DESCRIPTION
**Pull request by ktmud**
### SUMMARY

Encountered a broken unit test in `TimezoneSelector` after making changes in #19085 , found some room for improvements so I refactored this component a little: simplified the timezone overriding logics and made the unit tests more complete. See comments for details.

As the changes are quite significant and the component itself is small, it may be easier to review the diff in split view.

Xnip2022-03-09_17-32-11

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

No visual changes

### TESTING INSTRUCTIONS

Test places where timezone selector is used. The functionality should not change.

### ADDITIONAL INFORMATION

- [ ] Has associated issue: #15880 #17124
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/components/TimezoneSelector/index.tsx [hunk 1]
```diff
@@ -92,43 +92,40 @@ const TIMEZONE_OPTIONS = TIMEZONES.map(zone => ({
...
 }));
 
+const TIMEZONE_OPTIONS_SORT_COMPARATOR = (
+  a: typeof TIMEZONE_OPTIONS[number],
+  b: typeof TIMEZONE_OPTIONS[number],
+) =>
+  moment.tz(currentDate, a.timezoneName).utcOffset() -
+  moment.tz(currentDate, b.timezoneName).utcOffset();
+
+TIMEZONE_OPTIONS.sort(TIMEZONE_OPTIONS_SORT_COMPARATOR);
```
Comment (ktmud) [inline superset-frontend/src/components/TimezoneSelector/index.tsx:102]: #19085 requires select options to be pre-sorted.

File: superset-frontend/src/components/TimezoneSelector/TimezoneSelector.test.tsx [hunk 2]
```diff
@@ -18,60 +18,78 @@
...
+  expect(searchInput).toBeInTheDocument();
+
+  userEvent.click(searchInput);
+  const options = await getSelectOptions();
+  // select option is first
+  expect(options[0]).toHaveTextContent('GMT -05:00 (Eastern Standard Time)');
+  // others are ranked by offset
+  expect(options[1]).toHaveTextContent('GMT -11:00 (Pacific/Pago_Pago)');
+  expect(options[2]).toHaveTextContent('GMT -10:00 (Hawaii Standard Time)');
+  expect(options[3]).toHaveTextContent('GMT -10:00 (America/Adak)');
```
Comment (ktmud) [inline superset-frontend/src/components/TimezoneSelector/TimezoneSelector.test.tsx:79]: New tests for option ordering.

File: superset-frontend/src/components/TimezoneSelector/TimezoneSelector.test.tsx [hunk 3]
```diff
@@ -18,60 +18,78 @@
...
+
+  let findTitle = 'GMT -07:00 (Mountain Standard Time)';
+  if (isDaylight) {
+    findTitle = 'GMT -06:00 (Mountain Daylight Time)';
+  }
+  const selectOption = await screen.findByTitle(findTitle);
+  expect(selectOption).toBeInTheDocument();
+  userEvent.click(selectOption);
+  expect(onTimezoneChange).toHaveBeenCalledTimes(1);
+  expect(onTimezoneChange).toHaveBeenLastCalledWith('America/Cambridge_Bay');
```
Comment (ktmud) [inline superset-frontend/src/components/TimezoneSelector/TimezoneSelector.test.tsx:94]: Test the updated timezone value after selecting an option

File: superset-frontend/src/components/TimezoneSelector/index.tsx [hunk 4]
```diff
@@ -92,43 +92,40 @@ const TIMEZONE_OPTIONS = TIMEZONES.map(zone => ({
...
     }
-  }, [timezone, updateTimezone]);
+  }, [validTimezone, onTimezoneChange, timezone]);
 
   return (
     <Select
       ariaLabel={t('Timezone selector')}
       css={{ minWidth: MIN_SELECT_WIDTH }} // smallest size for current values
-      onChange={onTimezoneChange}
-      value={timezone || DEFAULT_TIMEZONE.value}
```
Comment (ktmud) [inline superset-frontend/src/components/TimezoneSelector/index.tsx:123]: Make sure `value` is a valid timezone. `timezone` may be different than `updatedTz`.

File: superset-frontend/src/components/TimezoneSelector/index.tsx [hunk 5]
```diff
@@ -92,43 +92,40 @@ const TIMEZONE_OPTIONS = TIMEZONES.map(zone => ({
...
-    (tz: string) => {
-      // update the ref to track changes
-      prevTimezone.current = tz;
-      // the parent component contains the state for the value
-      onTimezoneChange(tz);
-    },
-    [onTimezoneChange],
+  const validTimezone = useMemo(
+    () => matchTimezoneToOptions(timezone || moment.tz.guess()),
+    [timezone],
```
Comment (ktmud) [inline superset-frontend/src/components/TimezoneSelector/index.tsx:111]: Changed from `useRef` to simply triggering a setState in `useEffect`.

**Review by ktmud — COMMENTED** [review]
@eschutho could you confirm whether the updated code still produces the desired behavior for TimezoneSelector?

**Review by eschutho — APPROVED** [review]
Tested and it all works.

**Pull request merged**